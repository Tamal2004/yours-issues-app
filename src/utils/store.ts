import { Action } from 'redux';
import { combineReducers, configureStore, ThunkAction } from '@reduxjs/toolkit';
import { createWrapper } from 'next-redux-wrapper';

import * as issuesSlice from '../features/issues/issuesSlice';

export const rootReducer = combineReducers({
    [issuesSlice.name]: issuesSlice.reducer
});

const makeStore = () =>
    configureStore({
        reducer: rootReducer,
        devTools: true
    });

export type AppStore = ReturnType<typeof makeStore>;
export type RootState = ReturnType<AppStore['getState']>;
export type AppDispatch = AppStore['dispatch'];
export type AppThunk<ReturnType = void> = ThunkAction<
    ReturnType,
    RootState,
    unknown,
    Action
>;

export const wrapper = createWrapper<AppStore>(makeStore);
