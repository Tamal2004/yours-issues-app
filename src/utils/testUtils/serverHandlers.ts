import { rest } from 'msw';

import {
    mockedRepositoryNotFoundUrl,
    mockedRepositoryWithNoIssuesUrl,
    mockedRepositoryWithNoIssues,
    mockedRepositoryUrl,
    mockedRepository
} from 'services/__mocks__/repositoriesApiMocks';

import {
    mockedClosedIssuesCountUrl,
    mockedClosedIssuesCount,
    mockedGetOpenIssuesUrl,
    mockedGetOpenIssues,
    mockedGetClosedIssuesUrl,
    mockedGetClosedIssues
} from 'services/__mocks__/issuesApiMocks';

const handlers = [
    rest.get(mockedRepositoryNotFoundUrl, (_req, res, ctx) => {
        return res(ctx.status(403), ctx.delay(150));
    }),
    rest.get(mockedRepositoryWithNoIssuesUrl, (_req, res, ctx) => {
        return res(ctx.json(mockedRepositoryWithNoIssues), ctx.delay(150));
    }),
    rest.get(mockedRepositoryUrl, (_req, res, ctx) => {
        return res(ctx.json(mockedRepository), ctx.delay(150));
    }),
    rest.get(mockedClosedIssuesCountUrl, (_req, res, ctx) => {
        return res(ctx.json(mockedClosedIssuesCount), ctx.delay(150));
    }),
    rest.get(mockedGetOpenIssuesUrl, (_req, res, ctx) => {
        return res(ctx.json(mockedGetOpenIssues), ctx.delay(150));
    }),
    rest.get(mockedGetClosedIssuesUrl, (_req, res, ctx) => {
        return res(ctx.json(mockedGetClosedIssues), ctx.delay(150));
    })
];

export { handlers };
