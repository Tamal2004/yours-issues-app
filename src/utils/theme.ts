const composeFontSize = ({
    scaleMax = 100,
    scaleMin = 100,
    priorityWidth = undefined,
    precision = 3,
    widthMax,
    widthMin
}: {
    scaleMax?: number;
    scaleMin?: number;
    priorityWidth?: number;
    precision?: number;
    widthMax: number;
    widthMin: number;
}): string => {
    const baseScale = 25;
    const baseAbsoluteSize = 16;

    const medianWidth = priorityWidth
        ? priorityWidth
        : (widthMax - widthMin) / 2 + widthMin; // Mid way between widths

    const widthDiff = widthMax - widthMin;
    const median = ((medianWidth - widthMin) / widthDiff) * 100;

    const scaleDiff = (scaleMax - scaleMin) / 100; // Turn into decimal
    const relativeDiff = scaleDiff * baseScale;

    // The scale at the median mark. Is in percentage form (not decimal)
    const medianScaleDiff = (relativeDiff * median) / 100;

    // Needs to be converted to decimal equivalent of the percentage
    const absoluteMedianDiff = (medianScaleDiff * baseAbsoluteSize) / 100;

    // Convert median to decimal
    const medianWidthDiff = (widthDiff * median) / 100;

    /*
    Finding the intersect
   */

    // Width at which scaleDiff is 0
    const zeroGradientWidth = widthMax - medianWidthDiff;

    // Rise over run -- scale/width
    const gradient = absoluteMedianDiff / medianWidthDiff;

    // 0 = mx + b -- Flip into negative to get b (intersect)
    const intersect = -1 * gradient * zeroGradientWidth;

    // Gradient diff multiplied by 100vw -- taking the 100
    const viewWidthScale = `${(gradient * 100).toFixed(precision)}vw`;

    // Static scale, based on 16px -- will not scale with user-agent styles
    const staticScale = `(${viewWidthScale} + ${intersect.toFixed(
        precision
    )}px)`;

    // Accounts for the max scale being higher than 100%
    const absoluteMaxScale = (scaleMax / 100) * baseScale;

    // Dynamic scaling with user-agent font-size -- turns into a percentage
    const dynamicScale = `${(absoluteMaxScale - medianScaleDiff).toFixed(
        precision
    )}%`;

    return `calc(${dynamicScale} + ${staticScale})`;
};

export { composeFontSize };
