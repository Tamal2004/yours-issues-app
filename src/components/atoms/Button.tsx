import React from 'react';
import { css, styled } from 'theme';
import SpinningLoader from 'atoms/SpinningLoader';

type ButtonProps = JSX.IntrinsicElements['button'] & {
    loading?: boolean;
    className?: string; // CSS Overrides
};

const StyledButton = styled.button<{
    $disabled?: boolean;
    $loading?: boolean;
}>`
    position: relative;

    display: grid;
    justify-content: center;
    align-items: center;

    width: min-content;

    border-radius: 1rem;
    border: none;
    outline: none;

    cursor: pointer;
    white-space: nowrap;
    text-decoration: none;
    user-select: none;
    transition: all 300ms;
    background-color: ${({ theme }) => theme.colors.secondary.main};

    padding: 1.875rem 5.25rem;

    ${({ theme }) => theme.typography.button};

    ${({ $disabled, $loading }) => css`
        pointer-events: ${$disabled || $loading ? 'none' : 'initial'};
        opacity: ${$disabled || $loading ? '0.65' : '1'};
        color: ${$loading ? 'transparent' : 'white'};
    `}
`;

const LoaderWrapper = styled.div<{ $loading: boolean }>`
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    
    transition: all 300ms;
    opacity: ${({ $loading }) => $loading ? '1' : '0'};
`;

const Button: React.FC<ButtonProps> = ({
    disabled = false,
    loading = false,
    onClick,
    children,
    className,
    ref,
    ...restProps
}) => {
    return (
        <StyledButton
            className={className}
            $disabled={disabled}
            $loading={loading}
            disabled={disabled}
            onClick={onClick}
            {...restProps}
        >
            {children}
            {loading && (
                <LoaderWrapper $loading={loading}>
                    <SpinningLoader />
                </LoaderWrapper>
            )}
        </StyledButton>
    );
};

export default Button;
