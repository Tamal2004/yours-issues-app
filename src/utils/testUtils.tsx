import React from 'react';
import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';

import { AppStore, rootReducer, RootState } from 'utils/store';
import { theme } from 'theme';

const render = (
    ui: React.ReactElement,
    {
        preloadedState,
        store = configureStore({ reducer: rootReducer, preloadedState }),
        ...renderOptions
    }: { preloadedState?: RootState; store?: AppStore } & RenderOptions = {}
) => {
    const Wrapper: React.FC = ({ children }) => {
        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>{children}</ThemeProvider>
            </Provider>
        );
    };
    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};

// Re-export everything
export * from '@testing-library/react';
// Override render method
export { render };
