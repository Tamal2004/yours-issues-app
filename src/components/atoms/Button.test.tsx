import React from 'react';

import { render, fireEvent } from 'utils/testUtils';
import Button from './Button';

describe('<Button />', () => {
    it('should render a button', () => {
        const mockClick = jest.fn();
        const { getByText } = render(<Button onClick={mockClick}>Show</Button>);

        const TestButton = getByText('Show');
        expect(TestButton).toBeInTheDocument();
        expect(TestButton).toMatchSnapshot();
    });

    it('should perform the passed on click function', () => {
        const mockClick = jest.fn();
        const { getByText } = render(<Button onClick={mockClick}>Show</Button>);
        fireEvent.click(getByText('Show'));
        expect(mockClick.mock.calls.length).toBe(1);
    });

    it('should not perform passed on click function when disabled', () => {
        const mockClick = jest.fn();

        const { getByText } = render(
            <Button onClick={mockClick} disabled>
                Show
            </Button>
        );

        fireEvent.click(getByText('Show'));
        expect(mockClick.mock.calls.length).toBe(0);
    });

    it('should not perform passed on click function when loading', () => {
        const mockClick = jest.fn();

        const { getByText } = render(
            <Button onClick={mockClick} loading>
                Show
            </Button>
        );

        fireEvent.click(getByText('Show'));
        expect(mockClick.mock.calls.length).toBe(0);
    });
});
