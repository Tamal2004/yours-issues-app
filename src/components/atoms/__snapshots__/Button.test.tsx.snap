// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a button 1`] = `
<button
  class="sc-gsnTZi goDtiU"
  role="button"
>
  Show
</button>
`;
